lib/remap-functions/src/parse_key_value.rs

```diff
@@ -0,0 +1,351 @@
+    sequence::{delimited, preceded, tuple},
+    IResult,
+};
+use remap::prelude::*;
+use std::collections::BTreeMap;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseKeyValue;
+
+impl Function for ParseKeyValue {
+    fn identifier(&self) -> &'static str {
+        "parse_key_value"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: true,
+            },
+            Parameter {
+                keyword: "field_split",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+            Parameter {
+                keyword: "separator",
+                accepts: |v| matches!(v, Value::Bytes(_)),
+                required: false,
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+        let field_split = arguments.optional("field_split").map(Expr::boxed);
+        let separator = arguments.optional("separator").map(Expr::boxed);
+
+        Ok(Box::new(ParseKeyValueFn {
+            value,
+            field_split,
+            separator,
+        }))
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(crate) struct ParseKeyValueFn {
+    value: Box<dyn Expression>,
+    field_split: Option<Box<dyn Expression>>,
+    separator: Option<Box<dyn Expression>>,
+}
+
+impl Expression for ParseKeyValueFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let bytes = self.value.execute(state, object)?.try_bytes()?;
+        let value = String::from_utf8_lossy(&bytes);
+
+        let field_split = match &self.field_split {
+            Some(s) => String::from_utf8_lossy(&s.execute(state, object)?.try_bytes()?).to_string(),
+            None => "=".to_string(),
+        };
+
+        let separator = match &self.separator {
+            Some(s) => String::from_utf8_lossy(&s.execute(state, object)?.try_bytes()?).to_string(),
+            None => " ".to_string(),
+        };
+
+        let (_, values) =
+            parse_line(&value, &field_split, &separator).map_err(|e| e.to_string())?;
+
+        Ok(values.into_iter().collect::<BTreeMap<_, _>>().into())
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .merge_optional(self.field_split.as_ref().map(|field_split| {
+                field_split
+                    .type_def(state)
+                    .fallible_unless(value::Kind::Bytes)
+            }))
+            .merge_optional(self.separator.as_ref().map(|separator| {
+                separator
+                    .type_def(state)
+                    .fallible_unless(value::Kind::Bytes)
+            }))
+            .fallible_unless(value::Kind::Bytes)
```

Comment:
@JeanMertz I think this may be wrong. The function can fail even if the input is Bytes if it is not able to parse the input string correctly. So should this actually be : `.into_fallible(true)`?

I think there's a number of places in the code that do this..

Reply:
That's correct. If we can't know at compile-time if a function can be infallible, it should always be considered fallible.

That's usually the case with `parse_*` functions, such as `parse_json`. Technically, we _could_ check if the input is a string literal, and do the parsing at compile-time, but that's such a (mostly useless) edge-case that it's not worth pursuing right now.

This also means that once we enforce infallibility at runtime, operators are required to handle error cases for all(?) parse functions.